### PR TITLE
improve exception message if value doesn't match expected type

### PR DESCRIPTION
**Pull request by dirk-thomas**
Example command:

    ros2 topic pub -1 /goal geometry_msgs/PoseStamped '{header: {stamp: now, frame_id: "odom"}, pose: {position: {x: -2.0, y: 2.0, z: 0.0}, orientation: {w: 1.0}}}'

### CODE REVIEW

File: rosidl_runtime_py/set_message.py [hunk 1]
```diff
@@ -33,9 +33,15 @@ def set_message_fields(msg: Any, values: Dict[str, str]) -> None:
     :param values: The values to set in the ROS message. The keys of the dictionary represent
         fields of the message.
     :raises AttributeError: If the message does not have a field provided in the input dictionary.
-    :raises ValueError: If a message value does not match its field type.
+    :raises TypeError: If a message value does not match its field type.
```
Comment (dirk-thomas) [inline rosidl_runtime_py/set_message.py]: `ValueError` is incorrect here. If the value has an undesired type the correct exception type is `TypeError`. `ValueError` should only be used if the value type is correct but the value is e.g. outside a specific range.

File: rosidl_runtime_py/set_message.py [hunk 2]
```diff
@@ -33,9 +33,15 @@ def set_message_fields(msg: Any, values: Dict[str, str]) -> None:
...
     """
-    for field_name, field_value in values.items():
+    try:
+        items = values.items()
+    except AttributeError:
+        raise TypeError(
+            "Value '%s' is expected to be a dictionary but is a %s" %
```
Comment (clalancette) [inline rosidl_runtime_py/set_message.py]: I think we more specifically call these "fields", so:

```suggestion
            "Field '%s' is expected to be a dictionary but is a %s" %
```

Comment (dirk-thomas) [reply rosidl_runtime_py/set_message.py]: Since the message actually outputs the `values` the term "Field" isn't correct.

Comment (clalancette) [reply rosidl_runtime_py/set_message.py]: But what is in the '%s' *is* the field name.  The most correct is:

```
Value for field '%s' is expected to be a dictionary but is a %s.
```

It's a little wordy, but maybe that is better.

Comment (dirk-thomas) [reply rosidl_runtime_py/set_message.py]: No, it is not. When I run this patch for the above stated command I get the following error message:

> Value 'now' is expected to be a dictionary but is a str

And "not" is the value of the field "stamp".

Comment (clalancette) [reply rosidl_runtime_py/set_message.py]: Ah sorry, I got confused.  You are right, it is the actual value.

**Comment by dirk-thomas** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/9843/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/5459/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8005/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/9698/)

**Pull request merged**